### PR TITLE
Add `fn_to_numeric_cast_usize` lint

### LINKED ISSUE #878: Wrong hint about needless_range_loop
Clippy warns about following snippet:

```
let v = vec![5; 10];
for i in 0..v.len() - 2 {
    println!("{}", v[i + 2] - v[i]);
}
```

It suggest me to use `enumerate`. But I don't think I can use `enumerate` in this situation.

### LINKED ISSUE #5906: Check casting function item to `usize`
### What it does

Check a mistaken casting function item to `usize` or all other integer primitive types.

### Categories (optional)

- Kind: correctness or pedantic

*What is the advantage of the recommended code over the original code*

- It is a mistake to cast a function item to `usize` instead of getting function result.
  This could be unnotice for a long time.
- If you want to get pointer address of a function, explicitly cast it to function pointer and `usize`.
- Explicit is better than implicit.

The bug appeared in real code: https://github.com/rust-lang/rustup/commit/5d9d98059273c60de52c2abe573d79cd3f623de4
I was author of the bug :) .

### Drawbacks

It is verbose and somewhat redundant.

### Example

```rust
use std::mem::size_of;

extern "C" {
    fn foo(len: usize);
}

unsafe {
    foo(size_of::<usize> as usize);
}
```

Could be written as:

```rust
unsafe {
    foo(size_of::<usize>());
    // or
    foo(size_of::<usize> as fn() -> usize as usize);
    // or
    foo(size_of::<usize> as *const () as usize);
}
```

### PR DESCRIPTION
**Pull request by Coder-256**
changelog: Add `fn_to_numeric_cast_usize` lint

Fixes #5906.

In order to implement this, I had to make the design choice that casts from a function to `usize` should first go through a function pointer for clarity. For example:

```rust
fn fun() -> usize { 1 }
let a: usize = fun as fn() -> usize as usize;
```

Casts from a function *pointer* to a usize are still OK. Also, if you simply disable this lint, the old behavior should apply (except for error/help messages, as described below).

I also changed the error/help messages for "function as numeric" in order to more clearly explain that the function is being cast as a pointer, and suggest calling it then casting it, rather than casting a pointer to it as a `usize` (if I write `foo as i32`, I probably mean `foo() as i32`, not `foo as usize`).

Furthermore, I had to special-case the `transmutes_expressible_as_ptr_casts` lint in order to account for this change (recommending casting rather than calling due to the nature of transmute), and I am not 100% sure that I did it correctly. I based it on [this method](https://github.com/rust-lang/rust/blob/7e6d6e5f535321c2223f044caba16f97b825009c/src/librustc_middle/ty/mod.rs#L227-L241). This part might also need more tests.

Finally, this is my first contribution to Clippy, so please let me know if I missed anything.

---

EDIT: I've thought about this some more, and I think it would actually make more sense to remove the `fn_to_numeric_cast_with_truncation` lint entirely. I don't think the question of whether the cast would truncate is really relevant to this kind of lint; any cast to something other than `usize` is equally questionable IMO. Again, I think it's far more likely that the correct solution is to call the function and cast the result, rather than what the old help suggests: casting to a usize instead of whatever numeric were trying to cast to. I don't think I have ever needed to use a function pointer except for calling a C function.

Taking that into consideration, another solution could look something like:

| Original| Lint | Help |
|-|-|-|
| `func as i8` | `fn_to_numeric_cast` | `func() as i8`, [`func as fn() -> () as usize`] |
| `func as i128` | `fn_to_numeric_cast` | `func() as i128`, [`func as fn() -> () as usize`] |
| `func as usize` | `fn_to_numeric_cast_usize` | `func() as usize`, `func as fn() -> () as usize` |
| `fn_ptr as i8` | `fn_to_numeric_cast` | `fn_ptr() as i8`, [`fn_ptr as usize`] |
| `fn_ptr as i128` | `fn_to_numeric_cast` | `fn_ptr() as i128`, [`fn_ptr as usize`] |

The bracketed help messages are messages that I think would do more harm than good by being overly confusing (especially to new users) but should still be considered.

Casts that are allowed:
* `fn_ptr as usize`
* `fn_ptr as usize as i8`
* `fn_ptr as usize as i128`
* `func as fn() -> ()`
* `func as fn() -> () as usize`
* `func as fn() -> () as usize as i8`
* `func as fn() -> () as usize as i128`

I think I prefer this design better. You need to explicitly convert from a function to a function pointer, from a function pointer to a usize, and from a usize to another numeric. It's more verbose, but I think this way it is very hard to do the wrong thing by mistake. Plus, this way, the longer the cast, the less likely it is that you are doing the right thing! I'm not quite sure who makes these decisions though.

### CODE REVIEW

**Comment by Coder-256** [comment]
For the function item errors, I was also going to add something like "note: did you mean to cast to a function pointer?" but I'm not sure I know how to do that correctly. Should I just replace `span_lint_and_sugg` with `span_lint_and_then`, then cal `diag.span_suggestion` and `diag.help`?

**Comment by Coder-256** [comment]
Hmm, I had only guessed the 32-bit test output, and I just realized that I think I guessed wrong, but apparently CI doesn't test it(??), and the test failure is actually spurious(???).

Also, I implemented the alternate model in my [`fn-cast-usize-alternate branch`](https://github.com/Coder-256/rust-clippy/tree/fn-cast-usize-alternate), and I think the error messages are [even better than before](https://github.com/Coder-256/rust-clippy/compare/fn-cast-usize...Coder-256:fn-cast-usize-alternate#diff-997bf516f058eee2c0ee116a95f856a1R69-R83).

File: clippy_lints/src/transmute.rs [hunk 1]
```diff
@@ -630,6 +630,25 @@ impl<'tcx> LateLintPass<'tcx> for Transmute {
                             );
                         }
                     },
+                    (ty::FnDef(..), ty::Int(_) | ty::Uint(_)) => span_lint_and_then(
```
Comment (flip1995) [inline clippy_lints/src/transmute.rs:633]: ~~This is unrelated?~~

I wouldn't do this, since the extra cast to `fn() -> T` is code bloat. If people prefer these explicit casts, they should enable the new lint (which should be `pedantic`).

File: clippy_lints/src/types.rs [hunk 2]
```diff
@@ -1100,6 +1100,31 @@ declare_clippy_lint! {
...
+    /// fn fun() -> usize { 1 }
+    /// let a: usize = fun as usize;
+    ///
+    /// // Good
+    /// fn fun2() -> usize { 1 }
+    /// let a: usize = fun2 as fn() -> usize as usize;
+    /// let b: usize = fun2();
+    /// ```
+    pub FN_TO_NUMERIC_CAST_USIZE,
+    style,
```
Comment (flip1995) [inline clippy_lints/src/types.rs:1124]: This should be `pedantic` or even `restriction` IMO, since this cast is usually fine.

File: clippy_lints/src/types.rs [hunk 3]
```diff
@@ -1558,15 +1584,55 @@ fn lint_fn_to_numeric_cast(
...
+    let is_def = matches!(cast_from.kind, ty::FnDef(..));
+    let is_ptr = matches!(cast_from.kind, ty::FnPtr(_));
+
+    if is_def || is_ptr {
+        let mut applicability = Applicability::MaybeIncorrect;
+        let from_snippet = snippet_with_applicability(cx, cast_expr.span, "x", &mut applicability);
+
+        let to_nbits = int_ty_to_nbits(cast_to, cx.tcx);
+
+        if is_def {
```
Comment (flip1995) [inline clippy_lints/src/types.rs:1600]: This repeats too much code.

```rust
match cast_from.kind {
    ty::FnDef(..) if cast_to.kind == ty::Uint(UintTy::Usize) => { /* lint FN_TO_NUMERIC_CAST_USIZE */ },
    ty::FnDef(..) | ty::FnPtr(_) => { /* keep old code */ },
    .. => {},
}
```

File: tests/ui/fn_to_numeric_cast_32bit.rs [hunk 4]
```diff
@@ -19,10 +23,11 @@ fn test_function_to_numeric_cast() {
     let _ = foo as u32;
     let _ = foo as u64;
     let _ = foo as u128;
-
-    // Casting to usize is OK and should not warn
     let _ = foo as usize;
 
+    // Casting to fn pointer is OK and should not warn
+    let _ = foo as fn() -> String;
```
Comment (flip1995) [inline tests/ui/fn_to_numeric_cast_32bit.rs:29]: As an *additional* test: 
```rust
    // Casting to fn pointer and then to a `usize` is OK and should not warn
    let _ = foo as fn() -> String as usize;
```

File: clippy_lints/src/types.rs [hunk 5]
```diff
@@ -1558,15 +1584,55 @@ fn lint_fn_to_numeric_cast(
...
+                    applicability,
+                );
+            } else {
+                span_lint_and_sugg(
+                    cx,
+                    FN_TO_NUMERIC_CAST,
+                    expr.span,
+                    &format!("casting *pointer* to function `{}` as `{}`", from_snippet, cast_to),
+                    "try",
+                    format!("{}() as {}", from_snippet, cast_to),
```
Comment (flip1995) [inline clippy_lints/src/types.rs:1631]: I don't think we should assume here, that a cast of the return value was meant. Also I think the `casting function pointer ...` message was clear enough that it talks about a function pointer. Also emphasizing with * or something else isn't used in error messages.

What we can do is add two suggestions, one that suggests to cast the pointer to `usize` and one to call the function and cast the resulting value to `usize`, iff it is a `FnDef(..)`

**Comment by kraktus** [comment]
I think this can be closed as implemented by https://github.com/rust-lang/rust-clippy/pull/7705
